src/kernel/boot.c

```diff
@@ -874,20 +874,39 @@ BOOT_CODE bool_t init_freemem(word_t n_available, const p_region_t *available,
     word_t size = calculate_rootserver_size(it_v_reg, extra_bi_size_bits);
     word_t max = rootserver_max_size_bits(extra_bi_size_bits);
     for (; i >= 0; i--) {
-        word_t next = i + 1;
+        /* Invariant: both i and (i + 1) are valid indices in ndks_boot.freemem. */
+        assert(i < ARRAY_SIZE(ndks_boot.freemem) - 1);
+        /* Invariant; the region at index i is the current candidate.
+         * Invariant: regions 0 up to (i - 1), if any, are additional candidates.
+         * Invariant: region (i + 1) is undefined, and should not be read.
+         * Invariant: regions above (i + 1), if any, are empty or too small to use.
+         * Invariant: defined regions (all except i + 1) are ordered,
+         *            non-overlapping and unallocated. */
+
+        /* Can we fit a suitably aligned chunk in the top end of region i? */
```

Comment:
In my change I also added the check `if (reg->end - reg->start >= size) { ... ` here first, because blindly calculating `ndks_boot.freemem[i].end - size` could lead to an underflow and then we have a very large value that always looks right.